keel-core/src/main/kotlin/com/netflix/spinnaker/keel/persistence/Cleaner.kt

```diff
@@ -0,0 +1,96 @@
+package com.netflix.spinnaker.keel.persistence
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceId
+import com.netflix.spinnaker.keel.api.ResourceSpec
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.api.resources
+import com.netflix.spinnaker.keel.events.ResourceDeleted
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+import org.springframework.transaction.annotation.Propagation.REQUIRED
+import org.springframework.transaction.annotation.Transactional
+import java.time.Clock
+
+/**
+ * A class used to orchestrate the deletion of delivery configs
+ */
+@Component
+class Cleaner(
+  private val deliveryConfigRepository: DeliveryConfigRepository,
+  private val artifactRepository: ArtifactRepository,
+  private val resourceRepository: ResourceRepository,
+  private val clock: Clock,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  /**
+   * Deletes a delivery config and everything in it.
+   */
+  @Transactional(propagation = REQUIRED)
```

Comment:
Does this automagically take care of running all the methods that touch the DB in the same transaction?

Reply:
This might be a useful reference, BTW: https://stackoverflow.com/a/39998943

Reply:
Yes, it does. I've added some transactions to the other methods too